crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs

```diff
@@ -431,3 +441,141 @@ fn derive_libp2p_keypair(schnorrkel_sk: &schnorrkel::SecretKey) -> Keypair {
 
     Keypair::from(keypair)
 }
+
+/// Restores piece cache on startup. It waits for the new segment index and check all pieces from
+/// previous segments to see if they are already in the cache. If they are not, they are added
+/// from DSN.
+async fn restore_pieces_cache<PG, PC>(
+    mut run_cache_sync_rx: Receiver<SegmentIndex>,
+    piece_getter: Arc<FarmerPieceGetter<PG, PC>>,
+    piece_cache: Arc<tokio::sync::Mutex<FarmerPieceCache>>,
+) where
+    PG: PieceGetter + Send + Sync,
+    PC: PieceCache + Send + 'static,
+{
+    let piece_cache = piece_cache.clone();
+    let piece_getter = piece_getter.clone();
+
+    if let Some(segment_index) = run_cache_sync_rx.recv().await {
+        run_cache_sync_rx.close();
+
+        info!(new_segment_index=%segment_index, "Started syncing piece cache...");
+
+        for piece_index in 0..segment_index.first_piece_index().into() {
+            let key = PieceIndex::from(piece_index).hash().to_multihash().into();
+
+            let should_cache = { piece_cache.lock().await.should_cache(&key) };
```

Comment:
Just a though, it would have been faster to not take async lock every time, we could take a lock once and quickly skip through all first pieces that we already have, which should be most of them. We can even parallelize that (doing processing in batches of segment size).